src/java.base/share/classes/jdk/internal/foreign/CaptureStateUtil.java

```diff
@@ -0,0 +1,377 @@
+
+        // ((int|long))(int|long)
+        MethodHandle cleanup = MethodHandles.identity(returnType);
+        // (Throwable, (int|long))(int|long)
+        cleanup = MethodHandles.dropArguments(cleanup, 0, Throwable.class);
+        // (Throwable, (int|long), SegmentStack, C0=MemorySegment)(int|long)
+        // Cleanup does not have to have all parameters. It can have zero or more.
+        cleanup = MethodHandles.collectArguments(cleanup, 2, RELEASE_SEGMENT_MH);
+
+        // (SegmentStack, C0=MemorySegment, C1-Cn)(int|long)
+        MethodHandle tryFinally = MethodHandles.tryFinally(inner, cleanup);
+
+        // (SegmentStack, SegmentStack, C1-Cn)(int|long)
+        MethodHandle result = MethodHandles.filterArguments(tryFinally, 1, TAKE_SEGMENT_MH);
+
+        final MethodType newType = result.type().dropParameterTypes(0, 1);
+        perm = new int[result.type().parameterCount()];
+        perm[0] = 0;
+        for (int i = 1; i < result.type().parameterCount(); i++) {
+            perm[i] = i - 1;
+        }
+        // Deduplicate the first and second coordinate and only use the first
+        // (SegmentStack, C1-Cn)(int|long)
+        result = MethodHandles.permuteArguments(result, newType, perm);
+
+        // Finally we arrive at (C1-Cn)(int|long)
+        return MethodHandles.collectArguments(result, 0, SegmentPool.ACQUIRE_POOL_MH);
+    }
+
+    private static IllegalArgumentException illegalArgDoesNot(MethodHandle target, String info) {
+        return new IllegalArgumentException("The provided target " + target
+                + " does not " + info);
+    }
+
+    /**
+     * A cache of a memory segments to allow reuse. In many cases, only one segment will
+     * ever be allocated per platform thread. However, if a virtual thread becomes
+     * unmounted from its platform thread and another virtual thread acquires a new
+     * segment, there will be no reuse. Having a secondary cache of these is slower than
+     * using malloc/free directly.
+     * <p>
+     * The cache is about three times faster than a more general `deque`.
+     * <p>
+     * The class is using j.i.m.Unsafe rather than other supported APIs to allow early
+     * use in the boostrap sequence.
+     */
+    private static final class SegmentPool
+            extends SingleElementPool.SingleElementPoolImpl<MemorySegment> {
+
+        private static final TerminatingThreadLocal<SegmentPool> TL_POOLS =
+                new TerminatingThreadLocal<>() {
+                    @Override
+                    protected void threadTerminated(SegmentPool pool) {
+                        pool.close();
+                    }
+                };
+
+        private static final MethodHandle ACQUIRE_POOL_MH =
+                MhUtil.findStatic(MethodHandles.lookup(), "acquirePool",
+                        MethodType.methodType(SegmentPool.class));
+
+
+        private static final Supplier<MemorySegment> FACTORY = new Supplier<>() {
```

Comment:
I think turning factory and recycler into abstract methods in the superclass might lead to cleaner code